pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/customobject/DistinctTable.java

```diff
@@ -0,0 +1,297 @@
+        if (_sortedRecords.comparator().compare(record, leastRecord) > 0) {
+          _uniqueRecords.remove(leastRecord);
+          _uniqueRecords.add(record);
+          _sortedRecords.poll();
+          _sortedRecords.offer(record);
+        }
+      }
+      return true;
+    } else {
+      _uniqueRecords.add(record);
+      return _uniqueRecords.size() < _limit;
+    }
+  }
+
+  /**
+   * Returns {@code true} if no more records should be added, {@code false otherwise}. No more records should be added
+   * iff there is no order-by columns and enough unique records have been collected.
+   */
+  public boolean shouldNotAddMore() {
+    return _sortedRecords == null && _uniqueRecords.size() == _limit;
+  }
+
+  /*
+   * SERVER ONLY METHODS
+   */
+
+  /**
+   * (Server-side) Merges another {@code DistinctTable} into the current one.
+   * <p>NOTE: {@code DistinctTable} on Server-side has non-null {@code _uniqueRecords}.
+   */
+  public void serverSideMerge(DistinctTable distinctTable) {
+    if (shouldNotAddMore()) {
+      return;
+    }
+    for (Record record : distinctTable._uniqueRecords) {
+      if (!add(record)) {
+        return;
+      }
+    }
+  }
+
+  public byte[] toBytes()
+      throws IOException {
+    // NOTE: Serialize the DistinctTable as a DataTable
+    DataTableBuilder dataTableBuilder = new DataTableBuilder(_dataSchema);
+    DataSchema.ColumnDataType[] columnDataTypes = _dataSchema.getColumnDataTypes();
+    int numColumns = columnDataTypes.length;
+    for (Record record : _uniqueRecords) {
+      dataTableBuilder.startRow();
+      Object[] values = record.getValues();
+      for (int i = 0; i < numColumns; i++) {
+        switch (columnDataTypes[i]) {
+          case INT:
+            dataTableBuilder.setColumn(i, (int) values[i]);
+            break;
+          case LONG:
+            dataTableBuilder.setColumn(i, (long) values[i]);
+            break;
+          case FLOAT:
+            dataTableBuilder.setColumn(i, (float) values[i]);
+            break;
+          case DOUBLE:
+            dataTableBuilder.setColumn(i, (double) values[i]);
+            break;
+          case STRING:
+            dataTableBuilder.setColumn(i, (String) values[i]);
+            break;
+          case BYTES:
+            dataTableBuilder.setColumn(i, (ByteArray) values[i]);
+            break;
+          // Add other distinct column type supports here
+          default:
+            throw new IllegalStateException();
+        }
+      }
+      dataTableBuilder.finishRow();
+    }
+    return dataTableBuilder.build().toBytes();
+  }
+
+  /*
+   * BROKER ONLY METHODS
+   */
+
+  /**
+   * Broker-side constructor to deserialize the {@code DistinctTable} from a {@code ByteBuffer}. The
+   * {@code DistinctTable} constructed this way cannot be used to add records or merge other {@code DistinctTable}s
+   * but can only be merged into the main {@code DistinctTable} because the order-by and limit information is missing.
```

Comment:
I think calling it "main DistinctTable" sort of confuses it. What we should say is that:

- Server merges 1st to (N-1)th DistinctTable into 0th DistinctTable
- Broker creates a brand new empty DistinctTable (which will have the final records) and merges 0th to (N-1)th DistinctTable from each server into the new table. 

It looks like you are referring to the "new table" as the main table?

Reply:
There are 2 types of DistinctTables: Main DistinctTable and Deserialized DistinctTable. Added javadoc to explain them.